test/jaxp/javax/xml/jaxp/unittest/xpath/XPathPrecedingTest.java

```diff
@@ -0,0 +1,174 @@
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package xpath;
+
+import org.testng.Assert;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+import org.w3c.dom.Document;
+import org.w3c.dom.Node;
+import org.w3c.dom.NodeList;
+
+import javax.xml.parsers.DocumentBuilderFactory;
+import javax.xml.xpath.XPath;
+import javax.xml.xpath.XPathConstants;
+import javax.xml.xpath.XPathExpressionException;
+import javax.xml.xpath.XPathFactory;
+import java.io.ByteArrayInputStream;
+import java.io.InputStream;
+
+/*
+ * @test
+ * @bug 8289508
+ * @library /javax/xml/jaxp/unittest
+ *
+ * @run testng/othervm xpath.XPathPrecedingTest
+ * @summary Tests for XPath preceding and preceding-sibling axis specifiers.
+ */
+public class XPathPrecedingTest {
+
+    private static final String XML = """
+            <store>
+               <book id="1" lang="en">
+                  <title/>
+                  <author id="1"/>
+                  <isbn>1234</isbn>
+               </book>
+               <book id="2" lang="en">
+                  <title/>
+                  <author id="2"/>
+                  <isbn>5678</isbn>
+               </book>
+            </store>
+            """;
+    private static final Document doc;
+
+    static {
+        try {
+            var builder =
+                    DocumentBuilderFactory.newInstance().newDocumentBuilder();
+            InputStream s = new ByteArrayInputStream(XML.getBytes());
+            doc = builder.parse(s);
+        } catch (Exception e) {
+            System.out.println("Exception while initializing XML document");
+            throw new RuntimeException(e.getMessage());
+        }
+    }
+
+    /*
+     * DataProvider: provides XPath expression using preceding/preceding-sibling and the expected node(s) from the expression
+     */
+    @DataProvider(name = "preceding_axes")
+    public Object[][] getXPathPreceding() {
+        return new Object[][]{
+                // test preceding
+
+                // any nodes
+                {"/store/book[1]/author/preceding::*", "//book[1]/title"},
+
+                // abbreviated text
+                {"//author/preceding::title", "//title"},
```

Comment:
Note that both expressions return nodeset while the test calls for the 1st node. When performing node comparison, it would be good to select unique node.